fdbserver/workloads/ReportConflictingKeys.actor.cpp

```diff
@@ -0,0 +1,222 @@
+			endIdx = deterministicRandom()->randomInt(startPrefixIdx < endPrefixIdx ? 0 : startIdx, nodeCountPerPrefix);
+			startKey = keyForIndex(startPrefixIdx, startIdx);
+			endKey = keyForIndex(endPrefixIdx, endIdx);
+			tr->addWriteConflictRange(KeyRangeRef(startKey, endKey));
+		}
+	}
+
+	ACTOR Future<Void> conflictingClient(Database cx, ReportConflictingKeysWorkload* self, double delay,
+	                                     int actorIndex) {
+
+		state ReadYourWritesTransaction tr(cx);
+		state double lastTime = now();
+		state std::vector<KeyRange> readConflictRanges;
+
+		loop {
+			try {
+				// used for throttling
+				wait(poisson(&lastTime, delay));
+				if (self->reportConflictingKeys) tr.setOption(FDBTransactionOptions::REPORT_CONFLICTING_KEYS);
+				self->addRandomReadConflictRange(&tr, readConflictRanges);
+				self->addRandomWriteConflictRange(&tr);
+				++self->commits;
+				wait(tr.commit());
+				++self->xacts;
+			} catch (Error& e) {
+				TraceEvent("FailedToCommitTx").error(e);
+				state bool isConflict = false;
+				if (e.code() == error_code_operation_cancelled)
+					throw;
+				else if (e.code() == error_code_not_committed) {
+					++self->conflicts;
+					isConflict = true;
+				}
+				wait(tr.onError(e));
+				// check API correctness
+				if (!self->skipCorrectnessCheck && self->reportConflictingKeys && isConflict) {
+					const KeyRef conflictingKeysPreifx = LiteralStringRef("\xff\xff/transaction/conflicting_keys/");
+					state KeyRange ckr = KeyRangeRef(LiteralStringRef("").withPrefix(conflictingKeysPreifx),
+					                                 LiteralStringRef("\xff").withPrefix(conflictingKeysPreifx));
+					// The getRange here using the special key prefix "\xff\xff/transaction/conflicting_keys/" happens
+					// locally Thus, the error handling is not needed here
+					Future<Standalone<RangeResultRef>> conflictingKeyRangesFuture =
+					    tr.getRange(ckr, readConflictRanges.size() * 2);
+					ASSERT(conflictingKeyRangesFuture.isReady());
+					const Standalone<RangeResultRef> conflictingKeyRanges = conflictingKeyRangesFuture.get();
+					ASSERT(conflictingKeyRanges.size() && (conflictingKeyRanges.size() % 2 == 0));
+					for (int i = 0; i < conflictingKeyRanges.size(); i += 2) {
+						KeyValueRef startKeyWithPreifx = conflictingKeyRanges[i];
+						ASSERT(startKeyWithPreifx.value == conflictingKeysTrue);
+						KeyValueRef endKeyWithPrefix = conflictingKeyRanges[i + 1];
+						ASSERT(endKeyWithPrefix.value == conflictingKeysFalse);
+						// Remove the prefix of returning keys
+						Key startKey = startKeyWithPreifx.key.removePrefix(conflictingKeysPreifx);
+						Key endKey = endKeyWithPrefix.key.removePrefix(conflictingKeysPreifx);
+						KeyRangeRef kr = KeyRangeRef(startKey, endKey);
+						if (!std::any_of(readConflictRanges.begin(), readConflictRanges.end(), [&kr](KeyRange rCR) {
+							    // Read_conflict_range remains same in the resolver.
```

Comment:
This check is only determining if the reported read conflict range was one of the ones we added to the transaction, not whether the range was actually the cause of the conflict. I would suggest a more controlled test where you start two transactions with the same read version, commit one of them first, then commit the other. If the second commit does not fail, you can verify that the read conflict ranges of the second transactions do not overlap with the write conflict ranges of the first transaction. If the second commit does fail than you can check that all reported conflict ranges overlap with write conflict ranges from the first transaction.

Reply:
Look at the ConflictRange test, because it does something similar. You will also want to do the same thing that test does, and in the Test.txt file disable buggify and connection failures (connectionFailuresDisableDuration=100000
buggify=off)

Reply:
Done. It now does what you describe here.

>start two transactions with the same read version, commit one of them first, then commit the other. If the second commit does not fail, you can verify that the read conflict ranges of the second transactions do not overlap with the write conflict ranges of the first transaction. If the second commit does fail than you can check that all reported conflict ranges overlap with write conflict ranges from the first transaction.